### PR TITLE
parser: add register and frozen api to avoid data race

### LINKED ISSUE #29224: Data race in `TestCharsetIntroducer` in `parser/charset/charset.go`
https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_check_2/detail/tidb_ghpr_check_2/41184/pipeline

```
[2021-10-27T17:13:41.213Z] WARNING: DATA RACE
[2021-10-27T17:13:41.213Z] Read at 0x00c00025a600 by goroutine 99:
[2021-10-27T17:13:41.213Z]   runtime.mapaccess2_faststr()
[2021-10-27T17:13:41.213Z]       /usr/local/go/src/runtime/map_faststr.go:107 +0x0
[2021-10-27T17:13:41.213Z]   github.com/pingcap/tidb/parser/charset.GetCharsetInfo()
[2021-10-27T17:13:41.213Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/charset/charset.go:136 +0xbb
[2021-10-27T17:13:41.213Z]   github.com/pingcap/tidb/parser/charset.GetDefaultCollation()
[2021-10-27T17:13:41.213Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/charset/charset.go:122 +0x5e
[2021-10-27T17:13:41.213Z]   github.com/pingcap/tidb/parser/charset.GetDefaultCollationLegacy()
[2021-10-27T17:13:41.213Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/charset/charset.go:114 +0xc8
[2021-10-27T17:13:41.213Z]   github.com/pingcap/tidb/parser.yyParse()
[2021-10-27T17:13:41.213Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/parser.go:15109 +0x5c42e
[2021-10-27T17:13:41.213Z]   github.com/pingcap/tidb/parser.(*Parser).ParseSQL()
[2021-10-27T17:13:41.213Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/yy_parser.go:153 +0x353
[2021-10-27T17:13:41.213Z]   github.com/pingcap/tidb/parser.(*Parser).ParseOneStmt()
[2021-10-27T17:13:41.213Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/yy_parser.go:183 +0x12d
[2021-10-27T17:13:41.213Z]   github.com/pingcap/tidb/parser_test.RunRestoreTest()
[2021-10-27T17:13:41.213Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/parser_test.go:390 +0x6ee
[2021-10-27T17:13:41.213Z]   github.com/pingcap/tidb/parser_test.RunTest()
[2021-10-27T17:13:41.213Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/parser_test.go:371 +0x285
[2021-10-27T17:13:41.213Z]   github.com/pingcap/tidb/parser_test.TestLikeEscape()
[2021-10-27T17:13:41.213Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/parser_test.go:4558 +0x1a6
[2021-10-27T17:13:41.213Z]   testing.tRunner()
[2021-10-27T17:13:41.213Z]       /usr/local/go/src/testing/testing.go:1193 +0x202
[2021-10-27T17:13:41.214Z] 
[2021-10-27T17:13:41.214Z] Previous write at 0x00c00025a600 by goroutine 151:
[2021-10-27T17:13:41.214Z]   runtime.mapassign_faststr()
[2021-10-27T17:13:41.214Z]       /usr/local/go/src/runtime/map_faststr.go:202 +0x0
[2021-10-27T17:13:41.214Z]   github.com/pingcap/tidb/parser/charset.AddCharset()
[2021-10-27T17:13:41.214Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/charset/charset.go:467 +0x28b
[2021-10-27T17:13:41.214Z]   github.com/pingcap/tidb/parser_test.TestCharsetIntroducer()
[2021-10-27T17:13:41.214Z]       /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/parser/parser_test.go:6498 +0xd0
[2021-10-27T17:13:41.214Z]   testing.tRunner()
[2021-10-27T17:13:41.214Z]       /usr/local/go/src/testing/testing.go:1193 +0x202

```

https://github.com/pingcap/tidb/blob/18b47bcb2e9dd7919c78f0c5f59d435e1f6d5424/parser/charset/charset.go#L136

vs 

https://github.com/pingcap/tidb/blob/18b47bcb2e9dd7919c78f0c5f59d435e1f6d5424/parser/charset/charset.go#L466-L468

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #29224

Problem Summary: add register and frozen api to avoid data race

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: parser/charset/charset.go [hunk 1]
```diff
@@ -46,6 +47,7 @@ type Collation struct {
 	IsDefault   bool
 }
 
+var charsetRWLock sync.RWMutex
```
[the remark this answers is not in the document]
Comment (kennytm) [reply parser/charset/charset.go:50]: maybe just do the `frozen()` trick like pingcap/parser#1352.

[the remark this answers is not in the document]
Comment (xhebox) [reply parser/charset/charset.go:50]: I agree. Ping @Defined2014

[the remark this answers is not in the document]
Comment (Defined2014) [reply parser/charset/charset.go:50]: OK. I will try it later.

[the remark this answers is not in the document]
Comment (Defined2014) [reply parser/charset/charset.go]: Done. PTAL @xhebox @kennytm

File: parser/charset/charset.go [hunk 2]
```diff
@@ -49,6 +50,17 @@ type Collation struct {
 var collationsIDMap = make(map[int]*Collation)
 var collationsNameMap = make(map[string]*Collation)
 var supportedCollations = make([]*Collation, 0, len(supportedCollationNames))
+var registerFinish uint32
+
+// RegisterFinish makes the register of new error panic.
```
Comment (kennytm) [inline parser/charset/charset.go:55]:
```suggestion
// RegisterFinish makes the register of new charset panic.
```

**Review by kennytm — COMMENTED** [review]
Rest lgtm

File: tidb-server/main.go [hunk 3]
```diff
@@ -203,6 +204,8 @@ func main() {
 	// For regression of issue like https://github.com/pingcap/tidb/issues/28190
 	terror.RegisterFinish()
 
+	charset.RegisterFinish()
```
Comment (xiongjiwei) [inline tidb-server/main.go:207]: it still can not avoid data race in tests? the tests never run here

**Comment by xiongjiwei** [comment]
I have a proposal about charset in parser (https://internals.tidb.io/t/topic/456). It can solve the data race issue.

File: tidb-server/main.go (hunk 3 above)
Comment (Defined2014) [reply tidb-server/main.go:207]: Yes, tests couldn't run in parallel.

**Comment by Defined2014** [comment]
> I have a proposal about charset in parser (https://internals.tidb.io/t/topic/456). It can solve the data race issue totally

You mean clean up the code like `GetSupportedCharsets `? Maybe you can add some details about how to prevent data race.

**Comment by xiongjiwei** [comment]
it can not solve the data race actually, it still needs the test run serially

File: tidb-server/main.go (hunk 3 above)
Comment (xiongjiwei) [reply tidb-server/main.go:207]: then, this pr cannot close the issue

**Comment by Defined2014** [comment]
Close this PR, the test is run in serially, the data race will never happened in CI.

**Pull request closed without merging**